compiler/rustc_hir_typeck/src/fn_ctxt/suggestions.rs

```diff
@@ -704,10 +704,38 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
                 }
             }
             hir::FnRetTy::Return(ty) => {
+                let span = ty.span;
+
+                if let hir::TyKind::OpaqueDef(item_id, ..) = ty.kind
+                && let hir::Node::Item(hir::Item {
+                    kind: hir::ItemKind::OpaqueTy(op_ty),
+                    ..
+                }) = self.tcx.hir().get(item_id.hir_id())
+                && let hir::OpaqueTy {
+                    bounds: [bound], ..
+                } = op_ty
+                && let hir::GenericBound::LangItemTrait(
+                    hir::LangItem::Future, _, _, generic_args) = bound
+                && let hir::GenericArgs { bindings: [ty_binding], .. } = generic_args
+                && let hir::TypeBinding { kind, .. } = ty_binding
+                && let hir::TypeBindingKind::Equality { term } = kind
+                && let hir::Term::Ty(term_ty) = term {
+                    // Check if async function's return type was omitted.
+                    // Don't emit suggestions if the found type is `impl Future<...>`.
+                    debug!("suggest_missing_return_type: found = {:?}", found);
+                    if self.get_impl_future_output_ty(found).is_none() {
```

Comment:
can you use `is_suggestable` instead? https://doc.rust-lang.org/nightly/nightly-rustc/rustc_middle/ty/trait.IsSuggestable.html#tymethod.is_suggestable

Reply:
This seems to return false if I call `found.is_suggestable(self.tcx, false)`, so it doesn't emit the diagnostic if I use `is_suggestable` instead of `self.get_impl_future_output_ty(found).is_none()` unfortunately.